sdk/web-pubsub/web-pubsub-client/src/protocols/jsonProtocolBase.ts

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  AckMessage,
+  ConnectedMessage,
+  DisconnectedMessage,
+  GroupDataMessage,
+  JoinGroupMessage,
+  LeaveGroupMessage,
+  SendEventMessage,
+  SendToGroupMessage,
+  SequenceAckMessage,
+  ServerDataMessage,
+  WebPubSubDataType,
+  WebPubSubMessage,
+} from "../models/messages";
+import { JSONTypes } from "../webPubSubClient";
+
+export function parseMessages(input: string): WebPubSubMessage {
+  // The interface does allow "ArrayBuffer" to be passed in, but this implementation does not. So let's throw a useful error.
+  if (typeof input !== "string") {
+    throw new Error("Invalid input for JSON hub protocol. Expected a string.");
+  }
+
+  if (!input) {
+    throw new Error("No input");
+  }
+
+  const parsedMessage = JSON.parse(input);
+  const typedMessage = parsedMessage as { type: string; from: string; event: string };
+  let returnMessage: WebPubSubMessage;
+
+  if (typedMessage.type === "system") {
+    if (typedMessage.event === "connected") {
+      returnMessage = { ...parsedMessage, kind: "connected" } as ConnectedMessage;
+    } else if (typedMessage.event === "disconnected") {
+      returnMessage = { ...parsedMessage, kind: "disconnected" } as DisconnectedMessage;
+    } else {
+      throw new Error();
+    }
+  } else if (typedMessage.type === "message") {
+    if (typedMessage.from === "group") {
+      const data = parsePayload(parsedMessage.data, parsedMessage.dataType as WebPubSubDataType);
+      returnMessage = { ...parsedMessage, data: data, kind: "groupData" } as GroupDataMessage;
+    } else if (typedMessage.from === "server") {
+      const data = parsePayload(parsedMessage.data, parsedMessage.dataType as WebPubSubDataType);
+      returnMessage = {
+        ...parsedMessage,
+        data: data,
+        kind: "serverData",
+      } as ServerDataMessage;
+    } else {
+      throw new Error();
+    }
+  } else if (typedMessage.type === "ack") {
+    returnMessage = { ...parsedMessage, kind: "ack" } as AckMessage;
+  } else {
+    throw new Error();
+  }
+  return returnMessage;
+}
+
+export function writeMessage(message: WebPubSubMessage): string {
+  let data: any;
+  switch (message.kind) {
+    case "joinGroup": {
+      data = new JoinGroupData(message);
+      break;
+    }
+    case "leaveGroup": {
+      data = new LeaveGroupData(message);
+      break;
+    }
+    case "sendEvent": {
+      data = new SendEventData(message);
+      break;
+    }
+    case "sendToGroup": {
+      data = new SendToGroupData(message);
+      break;
+    }
+    case "sequenceAck": {
+      data = new SequenceAckData(message);
+      break;
+    }
+    default: {
+      throw new Error(`Unsupported type: ${message.kind}`);
+    }
+  }
+
+  return JSON.stringify(data);
+}
+
+class JoinGroupData {
```

Comment:
classes that simply hold data feel unnecessary to me, why not use an object that fits an interface?

Reply:
I think it not only hold data but has the conversion from data model used in sdk to wire protocol in constructor. I think put conversion together with classes make it clear